File: _cucumber/lib/helpers/site_user.rb
module SiteUser

  def generate_user(country=nil)

    country = ['United States', 'Czech Republic', 'United Kingdom', 'Canada', 'Mexico', 'Ukraine'].sample if country == nil

    case country
      when 'United States'
        state = 'Washington'
        city = 'Washington DC'
      when 'Canada'
        state = 'Ontario'
        city = 'Ottawa'
      when 'Mexico'
        state = 'Mexico'
        city = 'Mexico City'
      when 'Ukraine'
        state = nil
        city = 'Kiev'
      else
        state = nil
        city = nil
    end

    {
        email: "redhat-developers-testers+sid_#{$session_id}_#{Faker::Lorem.characters(5)}@redhat.com",
        greeting: %w(Mr. Mrs. Ms. Miss Dr. Hr Sr.).sample,
        first_name: Faker::Name.first_name,
        last_name: Faker::Name.last_name,
        password: Faker::Internet.password,
        company_name: Faker::Company.name,
        address_line_one: Faker::Address.street_address,
        city: city,
        state: state,
        postal_code: Faker::Address.postcode,
        country: country,
        phone_number: '0191 1111111'

    }

  end

  def github_account_holder
      {
          identity_provider: 'github',
          user_id: '20190656',
          user_name: 'rhdsociallogin',
          git_password: 'P@$$word01'
      }
  end

  def get_email(email_address)
    tries ||= 10
    begin
      @gmail = Gmail.connect!('uk.redhat.test01.user@gmail.com', '$tumpjumperRedH@tTe$t01')
    rescue => e
      if (tries -= 1) > 0
        puts "Failed to access gmail with exception: #{e}"
        sleep(1.5)
        retry
      else
        raise('Test user was not logged into gmail after 10 attempts!')
      end
    end

    try(7) {
      @email = @gmail.inbox.emails(:unread, to: email_address).last
    }

    message_body = @email.message.body.raw_source
    url = message_body.scan(/https?:\/\/[\S]+/).first
    # delete all mails and close the gmail session
    @gmail.inbox.find(:to => email_address).each do |email|
      email.delete!
    end
    @gmail.logout
    raise('Test user was not logged out of gmail session!') if @gmail.logged_in?.eql?(true)
    encoded_url = URI.encode(url)
    URI.parse(encoded_url)
    # return valid URI
    return encoded_url.gsub('<', '')
  end

  private

  def try(i)
    count = 0; email = nil
    until email != nil || count == i
      puts ' . . . waiting for email . . .'
      email = yield
      sleep 10
      count += 1
    end
    if email.eql?(nil)
      raise("Email was not received after #{count}0 seconds")
    end
  end

end

World(SiteUser)

File: _tests/e2e/keycloak/features/support/rest/Site.user.js
const faker = require('faker');
const KeyCloakAdmin = require("./Keycloak.admin");
const ITAdmin = require("./IT.admin");
const GetMail = require('./Gmail');

class SiteUser {

    generate(country = null) {
        return this._generateCustomerCredentials(country)
    }

    createRHDSiteUser() {
        let user = this.generate();
        let keycloakAdmin = new KeyCloakAdmin();
        user['rhUserId'] = keycloakAdmin.registerNewUser(user);
        // return registered user details, and rhd User ID
        return user
    }

    createOpenshiftUser() {
        let user = this.generate();
        let itAdmin = new ITAdmin();
        itAdmin.createSimpleUser(user);
        // return registered user details
        return user
    }

    createCustomerPortalAccount() {
        let user = this.generate();
        let itAdmin = new ITAdmin();
        itAdmin.createFullUser(user);
        // return registered user details
        return user
    }

    disableCustomerPortalAccount(user) {
        let itAdmin = new ITAdmin();
        itAdmin.disableUser(user);
    }

    verifyRHDAccount(email) {
        let gmail = new GetMail();
        return gmail.process(email);
    }

    createUserWithLinkedSocialAccount() {
        let user = this.generate();
        let keycloakAdmin = new KeyCloakAdmin();
        keycloakAdmin.registerNewUser(user);
        keycloakAdmin.linkSocialProvider(user['email'], 'github', '20190656', 'rhdsociallogin');
        // return registered user details
        return user
    }

    getUserAttribute(email, attribute) {
        let keycloakAdmin = new KeyCloakAdmin();
        return keycloakAdmin.getUserAttribute(email, attribute)
    }


    /**
     private functions

     /**
     * Generate unique customer credentials
     * @private
     * @param  {Object} country
     * @return {Hash}: Customer credentials
     */
    _generateCustomerCredentials(country) {
        let selectCountry;
        let details;
        let extraAddressArgs;

        if (country === null) {
            selectCountry = this._selectRandom(['United Kingdom', 'United States'])
        } else {
            selectCountry = country
        }

        extraAddressArgs = this._getExtraAddressArgs(selectCountry);

        let emailAddress = `redhat-developers-testers+sid_${process.env.SESSION_ID}_${faker.random.number()}@redhat.com`;

        details = {
            title: this._selectRandom(['Mr', 'Mrs']),
            firstName: faker.name.firstName(),
            lastName: faker.name.lastName(),
            email: emailAddress,
            username: emailAddress.replace('@redhat.com', ''),
            password: 'P@$$word01',
            company: faker.company.companyName(),
            phoneNumber: faker.phone.phoneNumber(),
            country: selectCountry,
            addressLineOne: faker.address.streetName(),
            city: extraAddressArgs['city'],
            state: extraAddressArgs['state'],
            postalCode: faker.address.zipCode(),
            countryCode: extraAddressArgs['countryCode']
        };
        return details
    }


    /**
     * Get US city and state if country selected is US
     * @private
     * @param  {Object} country
     * @return {Hash}: US city and state, otherwise return null
     */
    _getExtraAddressArgs(country) {
        let details;
        if (country === 'United States') {
            details = {
                state: 'Washington',
                city: 'Washington DC',
                countryCode: 'US'
            };
        } else {
            details = {
                state: null,
                city: null,
                countryCode: 'GB'
            };
        }
        return details;
    }

    _selectRandom(array) {
        return array[Math.floor(Math.random() * array.length)];
    }
}

module.exports = SiteUser;
